community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java

```diff
@@ -559,14 +579,28 @@ public void transactionRecovered( long txId )
                     transactionLogModule.logFiles(), transactionLogModule.storeFlusher(), startupStatistics,
                     legacyIndexApplierLookup );
 
-            KernelModule kernelModule = buildKernel( indexingModule.integrityValidator(),
+            final KernelModule kernelModule = buildKernel( indexingModule.integrityValidator(),
                     transactionLogModule.transactionAppender(), neoStoreModule.neoStores(),
                     transactionLogModule.storeApplier(), indexingModule.indexingService(),
                     indexingModule.indexUpdatesValidator(),
                     storeLayerModule.storeLayer(),
                     cacheModule.updateableSchemaState(), indexingModule.labelScanStore(),
                     indexingModule.schemaIndexProviderMap(), cacheModule.procedureCache() );
 
+            if ( safeIdBuffering )
+            {
+                // Now that we've instantiated the component which can keep track of transaction boundaries
+                // we let the id generator know about it.
+                bufferingIdGeneratorFactory.initialize( new Supplier<BooleanSupplier>()
```

Comment:
Supplier of supplier could get a bit simpler if `KernelTransactionsSnapshot` is used here.